source/common/http/header_map_impl.h

```diff
@@ -256,6 +291,176 @@ class HeaderMapImpl : NonCopyable {
+
+  // Http::HeaderMap
+  bool operator==(const HeaderMap& rhs) const override;
+  bool operator!=(const HeaderMap& rhs) const override;
+  void addViaMove(HeaderString&& key, HeaderString&& value) override;
+  void addReference(const LowerCaseString& key, absl::string_view value) override;
+  void addReferenceKey(const LowerCaseString& key, uint64_t value) override;
+  void addReferenceKey(const LowerCaseString& key, absl::string_view value) override;
+  void addCopy(const LowerCaseString& key, uint64_t value) override;
+  void addCopy(const LowerCaseString& key, absl::string_view value) override;
+  void appendCopy(const LowerCaseString& key, absl::string_view value) override;
+  void setReference(const LowerCaseString& key, absl::string_view value) override;
+  void setReferenceKey(const LowerCaseString& key, absl::string_view value) override;
+  void setCopy(const LowerCaseString& key, absl::string_view value) override;
+  uint64_t byteSize() const override;
+  const HeaderEntry* get(const LowerCaseString& key) const override;
+  void iterate(ConstIterateCb cb, void* context) const override;
+  void iterateReverse(ConstIterateCb cb, void* context) const override;
+  Lookup lookup(const LowerCaseString& key, const HeaderEntry** entry) const override;
+  void clear() override;
+  size_t remove(const LowerCaseString& key) override;
+  size_t removePrefix(const LowerCaseString& key) override;
+  size_t size() const override { return all_headers_num_; }
+  bool empty() const override { return all_headers_num_ == 0; }
+  void dumpState(std::ostream& os, int indent_level = 0) const override;
+  bool canCoalesce(const HeaderString& key) const {
+    // At the moment only inline-headers are coalesced.
+    return isInlineHeader(key.getStringView());
+  }
+  virtual bool isInlineHeader(absl::string_view key) const PURE;
+
+protected:
+  struct HeaderEntryImpl : public HeaderEntry, NonCopyable {
+    HeaderEntryImpl(const LowerCaseString& key);
+    HeaderEntryImpl(const LowerCaseString& key, HeaderString&& value, uint32_t index,
+                    bool coalesced);
+    HeaderEntryImpl(HeaderString&& key, HeaderString&& value, uint32_t index, bool coalesced);
+
+    // HeaderEntry
+    const HeaderString& key() const override { return key_; }
+    void value(absl::string_view value) override;
+    void value(uint64_t value) override;
+    void value(const HeaderEntry& header) override;
+    const HeaderString& value() const override { return value_; }
+    HeaderString& value() override { return value_; }
+
+    HeaderString key_;
+    HeaderString value_;
+    // Used to indicate the logical index of the header entry in the map, and
+    // to determine the order of header entries when iterating over the map.
+    // Note that for iteration, the indices must be different, but not
+    // consecutive (i.e., when removing a header)
+    uint32_t index_;
+    bool coalesced_;
+  };
+
+  using HeaderEagerMapValue = std::list<HeaderEntryImpl>;
```

Comment:
I would recommend using InlinedVector here of size 1. This will avoid an extra allocation in the common case, and given that we don't allow deletion of individual non-coalesced headers, there will never be reshuffle during deletion.

Reply:
That's a good idea, and that was the first thing I attempted.
The problem was that `HeaderEntryImpl` is non-copyable, and if it is made copyable, need to introduce some changes to HeaderString. It might make sense to do this, just hesitant introducing more changes than probably needed.

Reply:
yeah good point re: the copying, though, if you end up with a node hash map (or flat hash map with unique_ptr values) I think you can avoid this as a side benefit.

Reply:
I think you are correct, I'll test it out (another comparison point), thanks!